File: BinaryTreeRightView.java
/*
// Time Complexity : O(N)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :No


// Your code here along with comments explaining your approach
/*
We will use Queue to store tree nodes .We push root node to queue.Pop the root node and add its left and right node
to queue.Keeping processing the nodes until queue is empty 
*/


import java.util.*;

import javax.swing.tree.TreeNode;
public class BinaryTreeRightView {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> output = new ArrayList<>();

        if(root == null) return output;

        Queue<TreeNode> queue = new LinkedList<>();

        queue.add(root);

        while(!queue.isEmpty()){
            int size = queue.size();

            for(int index = 1;  index <= size; index++){
                TreeNode node = queue.remove();

                if(index  == size){
                    output.add(node.val);
                }

                if(node.left != null){
                    queue.add(node.left);
                }
                if(node.right != null){
                    queue.add(node.right);
                }
            }
        }
        return output;
    }
}

File: binarytreeRightView.py
# Time Complexity : O(N)
# Space Complexity : O(N)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root) :
        result = []
        if root is None:
            return result
        queue = deque()
        queue.append(root)  # store elements visited in each level

        while queue:
            last_node = None  # last node at each level
            for i in range(len(queue)):
                node = queue.popleft()
                last_node = node

                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)

            result.append(last_node.val)  # append the last node to the result

        return result
